server/server.js

```diff
@@ -157,7 +167,75 @@ function handleRequest(request, response) {
           response.end(data);
         });
       break;
+    case "image-upload":
+        var boardName = parts[1];
+        if (!boardName) {
+          response.writeHead(400, { "Content-Type": "text/plain" });
+          response.end("No board name provided");
+          return;
+        }
+
+        const { fields, files } = await parseFormData(request);
+        if (!files?.image[0]) {
+          response.writeHead(400, { "Content-Type": "text/plain" });
+          response.end("No image provided");
+          return;
+        }
+
+        try {
+          validateImage(files.image[0]);
+        } catch (error) {
+          response.writeHead(400, { "Content-Type": "text/plain" });
+          response.end(error.message);
+          return;
+        }
+
+        const id = fields.id[0];
+        const position = JSON.parse(fields.position[0]);
+        const dimensions = JSON.parse(fields.dimensions[0]);
+        const image = files.image[0];
+
+        if (!boardDataList.has(boardName)) {
+          response.writeHead(404, { "Content-Type": "text/plain" });
+          response.end("Board not found");
+          return;
+        }
+
+        // Save the image to disk.
+        const board = await boardDataList.get(boardName);
+        const { publicAssetPath, } = await board.saveImageAsset(id, image)
+
+        // Notify all clients of the new image.
+        socketsInstance.handleImageUpload(
+          boardName, id, publicAssetPath, position, dimensions
+        );
 
+        // Upload successful!
+        response.writeHead(200, {
+          "Content-Type": "application/json",
+          "Content-Security-Policy": CSP,
+          "Cache-Control": "public, max-age=30",
+        });
+      response.end(JSON.stringify({ status: 'ok' }));
+      break;
+    case "board-assets":
+        const [, boardId, assetName] = parts;
+        const file = fs.readFileSync(path.join(config.HISTORY_DIR, `board-${boardId}`, assetName));
```

Comment:
also, we cannot have blocking code like this in the request handling function

Reply:
I can do some more research on this but do you happen to know if calling a function that returns a promise would solve this issue?

Something like this:

```js
async function readFile(request, response) {
  // ... create promise, read file from disk, and handle response.
}

function handleRequest(request, response) {
  // ... handlers
  case 'board-assets':
     serveFile(request, response);
     break;
  // ... other handlers
}
```

Reply:
you can look at how it's done in the rest on the code; we use async fs functions